### PR TITLE
Enable LVI mitigation

### PR DESCRIPTION
**Pull request by mingweishih**
This PR supports the LVI mitigation (based on Intel's toolchain) in OE.
The summary of the changes
- Add `LVI_MITIGATION` and `LVI_MITIGATION_BINDIR` cmake options.
- Support the LVI_MITIGATION=ControlFlow build option, which generates
  the LVI-mitigated version of enclave libraries.
- Modify the `helloworld` sample to demonstrate how to build enclaves
  against LVI-mitigated libraries.
- Add tests with LVI mitigation.
- Add documentation.

Signed-off-by: Ming-Wei Shih <mishih@microsoft.com>

### CODE REVIEW

**Comment by mingweishih** [comment]
Will wait until CI/CD changes is done so we can run the test.

**Comment by radhikaj** [comment]
Bors try

File: 3rdparty/mbedtls/CMakeLists.txt [hunk 1]
```diff
@@ -155,39 +155,42 @@ maybe_build_using_clangw(mbedtls)
 
 # Change the mbedcrypto_static target to the real mbedcrypto library name.
 set_target_properties(mbedcrypto_static PROPERTIES OUTPUT_NAME "mbedcrypto")
+if (LVI_MITIGATION MATCHES ControlFlow)
+  set_target_properties(mbedcrypto_static-lvi-cfg PROPERTIES OUTPUT_NAME "mbedcrypto-lvi-cfg")
```
Comment (tomtau) [inline 3rdparty/mbedtls/CMakeLists.txt]: Does this work with the inline assembly -- e.g. https://github.com/openenclave/openenclave/blob/master/3rdparty/mbedtls/mbedtls/library/aesni.c#L125 ?

Comment (mingweishih) [reply 3rdparty/mbedtls/CMakeLists.txt]: Yes, this works with inline assembly. The instrumentation is done by the assembler.

**Comment by mingweishih** [comment]
Bors try

File: 3rdparty/mbedtls/CMakeLists.txt (hunk 1 above)
Comment (mingweishih) [reply 3rdparty/mbedtls/CMakeLists.txt]: To be more precise, the mitigation that this PR enables supports inline assembly. However, the PR only supports the type of "Control-Flow-Mitigation" that the Intel's deep dive article [1] describles. That is to say, the supported mitigation does not take care of all the load instructions such as the one in your example. Instead, it focuses on control-flow pattern such as indirect branches and return. Please let me know you have more question.

[1] https://software.intel.com/security-software-guidance/insights/deep-dive-load-value-injection#applyenclavemitigation

**Comment by mingweishih** [comment]
Fixed #2655

**Comment by ionutbalutoiu** [comment]
bors try

File: tests/crypto/enclave/CMakeLists.txt [hunk 2]
```diff
@@ -6,4 +6,4 @@ if (BUILD_ENCLAVES)
 endif()
 
 add_subdirectory(host)
-add_enclave_test(tests/crypto/enclave cryptohost cryptoenc)
+add_enclave_test(tests/crypto/enclave cryptohost crypto_enc)
```
Comment (radhikaj) [inline tests/crypto/enclave/CMakeLists.txt]: Why the change to crypto_enc

File: 3rdparty/mbedtls/CMakeLists.txt (hunk 1 above)
Comment (tomtau) [reply 3rdparty/mbedtls/CMakeLists.txt]: @mingweishih Will some of the cryptographic code need additional manual patching? The second case study in https://lviattack.eu/lvi.pdf has a similar example, but takes some extra assumptions for demonstration (known ciphertext, side-channel...)

Comment (mingweishih) [reply 3rdparty/mbedtls/CMakeLists.txt]: We currently don't have plan beyond the support of control-flow mitigation given that the cost of full mitigation is too high. If there is concern that control-flow mitigation is not sufficient, we will also consider adding further support.

**Pull request merged**